presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java

```diff
@@ -49,33 +49,53 @@ public OrderableArrayColumnValidator()
         Type elementType = ((ArrayType) column.getType()).getElementType();
 
-        FunctionCall arraySort = new FunctionCall(QualifiedName.of("array_sort"), ImmutableList.of(column.getIdentifier()));
         Expression checksum;
-        if (elementType instanceof ArrayType || elementType instanceof RowType) {
-            checksum = new CoalesceExpression(
-                    new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(new TryExpression(arraySort))),
-                    new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(column.getIdentifier())));
+        if (elementType.isOrderable()) {
+            FunctionCall arraySort = new FunctionCall(QualifiedName.of("array_sort"), ImmutableList.of(column.getIdentifier()));
+            if (elementType instanceof ArrayType || elementType instanceof RowType) {
+                checksum = new CoalesceExpression(
+                        new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(new TryExpression(arraySort))),
+                        new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(column.getIdentifier())));
+            }
+            else {
+                checksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(arraySort));
+            }
         }
         else {
-            checksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(arraySort));
+            checksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(column.getIdentifier()));
         }
 
-        return ImmutableList.of(new SingleColumn(checksum, Optional.of(delimitedIdentifier(getChecksumColumnAlias(column)))));
+        FunctionCall arrayCardinality = new FunctionCall(QualifiedName.of("sum"),
+                ImmutableList.of(new FunctionCall(QualifiedName.of("cardinality"), ImmutableList.of(column.getIdentifier()))));
+
+        return ImmutableList.of(new SingleColumn(checksum, Optional.of(delimitedIdentifier(getChecksumColumnAlias(column)))),
+                new SingleColumn(arrayCardinality, Optional.of(delimitedIdentifier(getCardinalityColumnAlias(column)))));
     }
 
     @Override
     public ColumnMatchResult validate(Column column, ChecksumResult controlResult, ChecksumResult testResult)
     {
-        String sortedChecksumColumnAlias = getChecksumColumnAlias(column);
-        Object controlSortedChecksum = controlResult.getChecksum(sortedChecksumColumnAlias);
-        Object testSortedChecksum = testResult.getChecksum(sortedChecksumColumnAlias);
+        String checksumColumnAlias = getChecksumColumnAlias(column);
+        Object controlChecksum = controlResult.getChecksum(checksumColumnAlias);
+        Object testChecksum = testResult.getChecksum(checksumColumnAlias);
+
+        String cardinalityColumnAlias = getCardinalityColumnAlias(column);
+        long controlCardinalitySum = (long) controlResult.getChecksum(cardinalityColumnAlias);
```

Comment:
@ShenYi This can may result in NullPointerException. When all arrays are null, sum(cardinality(a)) will be null and casting null to long will throw NPE. I'd drop this cast altogether.

```
Object controlCardinalitySum = controlResult.getChecksum(cardinalityColumnAlias);
Object testCardinalitySum = testResult.getChecksum(cardinalityColumnAlias);
```

Reply:
Nice catch. Fixed.